bugwarrior/services/linear.py

```diff
@@ -0,0 +1,181 @@
+import re
+import typing
+import json
+
+import requests
+
+from bugwarrior import config
+from bugwarrior.services import Service, Issue, Client
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class LinearConfig(config.ServiceConfig):
+    service: typing.Literal["linear"]
+    api_token: str
+
+    host: config.StrippedTrailingSlashUrl = config.StrippedTrailingSlashUrl(
+        "https://api.linear.app/graphql", scheme="https", host="api.linear.app"
+    )
+    import_labels_as_tags: bool = False
+    label_template: str = "{{label|replace(' ', '_')}}"
+
+
+class LinearIssue(Issue):
+    URL = "linearurl"
+    TITLE = "lineartitle"
+    DESCRIPTION = "lineardescription"
+    STATE = "linearstate"
+    IDENTIFIER = "linearidentifier"
+    TEAM = "linearteam"
+    CREATOR = "linearcreator"
+    ASSIGNEE = "linearassignee"
+    CREATED_AT = "linearcreated"
+    UPDATED_AT = "linearupdated"
+    CLOSED_AT = "linearclosed"
+
+    UDAS = {
+        URL: {"type": "string", "label": "Issue URL"},
+        TITLE: {"type": "string", "label": "Issue Title"},
+        DESCRIPTION: {"type": "string", "label": "Issue Description"},
+        STATE: {"type": "string", "label": "Issue State"},
+        IDENTIFIER: {"type": "string", "label": "Linear Identifier"},
+        TEAM: {"type": "string", "label": "Project ID"},
+        CREATOR: {"type": "string", "label": "Issue Creator"},
+        ASSIGNEE: {"type": "string", "label": "Issue Assignee"},
+        CREATED_AT: {"type": "date", "label": "Issue Created"},
+        UPDATED_AT: {"type": "date", "label": "Issue Updated"},
+        CLOSED_AT: {"type": "date", "label": "Issue Closed"},
+    }
+
+    UNIQUE_KEY = (URL,)
+
+    def parse_date(self, input):
+        '''Parse a date, stripping microseconds'''
+        parsed = super().parse_date(input)
+        if parsed:
+            parsed = parsed.replace(microsecond=0)
+        return parsed
+
+    def to_taskwarrior(self):
+        description = self.record.get("description")
+        created = self.parse_date(self.record.get("createdAt"))
+        modified = self.parse_date(self.record.get("updatedAt"))
+        closed = self.parse_date(self.record.get("completedAt"))
+
+        # Get a value, defaulting empty results to the given default.
+        def get(v, k, default=None):
+            r = v.get(k, default)
+            if not r:
+                return default
+            return r
+
+        return {
+            "project": re.sub(
+                r"[^a-zA-Z0-9]", "_", get(get(self.record, "project", {}), "name", "")
```

Comment:
This is an empty string since it's an argument to `re.sub`. That came from PivotalTracker, but it looks like project is always set there. I'll experiment a little more.